### ISSUE TITLE
operator "mod" and "%" evaluates to the same result

### ISSUE THREAD

**Issue by dynamicboy** [VBAndCs/sVB-Small-Visual-Basic#35]
Verson: Small Visual Basic v2.8.7
Details: 
the operator "mod" is supposed to get the quotient
the operator "%" is supposed to get the remainder

But I get  the same result with hat "mod" and "%" .

`Small Visual Basic`
Evaluator.Expression = "x % 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

Evaluator.Expression = "x mod 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

[CodeScreenShot.png](https://postimg.cc/FdxkrDGQ)

[ResultScreenShot.png](https://postimg.cc/PC66xQ3Y)

**Comment by VBAndCs** [comment]
> the operator "mod" is supposed to get the quotient



No. The Mod operator in VB, VB.NET, SB, and sVB returns the remainder. The % does the same in non BASIC languages like C++, C#, Python… etc, so I decided to add it as an alternative option in the math evaluator (but not in sVB code).

The \ operator (not /) is the integral operator in VB and VB.NET, but it is not supported in SB nor sVB. Instead you should use the Math methods Like Math.Floor(), Math.Ceiling() or math.Round() to get the integer part. I realized now there is no Math.Truncate() method, but you can get its result by calling Math.Floor(Math.Abs(x)). In the Evaluator you can just write:

Floor(Abs(x/3)) which will give you the result of x\3.

I will add the missing Truncate method, and I may consider adding the missing integral operator to sVB too.

Thanks.





________________________________
From: dynamicboy ***@***.***>
Sent: Wednesday, November 29, 2023 9:34 PM
To: VBAndCs/sVB-Small-Visual-Basic ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [VBAndCs/sVB-Small-Visual-Basic] operator "mod" and "%" evaluates to the same result (Issue #35)


Verson: Small Visual Basic v2.8.7
Details:
the operator "mod" is supposed to get the quotient
the operator "%" is supposed to get the remainder

But I get the same result with hat "mod" and "%" .

Small Visual Basic
Evaluator.Expression = "x % 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

Evaluator.Expression = "x mod 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

CodeScreenShot.png<https://postimg.cc/FdxkrDGQ>

ResultScreenShot.png<https://postimg.cc/PC66xQ3Y>

**Comment by VBAndCs** [comment]
Correction:

Floor(abs()) will change thee sign of the negative number. Seem that no alternative for Truncate.

________________________________
From: dynamicboy ***@***.***>
Sent: Wednesday, November 29, 2023 9:34 PM
To: VBAndCs/sVB-Small-Visual-Basic ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [VBAndCs/sVB-Small-Visual-Basic] operator "mod" and "%" evaluates to the same result (Issue #35)


Verson: Small Visual Basic v2.8.7
Details:
the operator "mod" is supposed to get the quotient
the operator "%" is supposed to get the remainder

But I get the same result with hat "mod" and "%" .

Small Visual Basic
Evaluator.Expression = "x % 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

Evaluator.Expression = "x mod 10"
TW.WriteLine(Evaluator.Evaluate(87)) ' output is 7

CodeScreenShot.png<https://postimg.cc/FdxkrDGQ>

ResultScreenShot.png<https://postimg.cc/PC66xQ3Y>

**Comment by dynamicboy** [comment]
@VBAndCs Thanks you. 
I used "Math.Floor( dividend / 10)" to get the integral part  as an alternative.
I just came across to this by trying to convert an integer to a string with a given width （12 -> 0012），as the "Text" library doesn't have such functions.
I wrote one as below：

``` Small Visual Basic

Function IntToStr(dbl_num, dbl_width)
   resultStr = ""
   
   digits = {}
   
   dividend = dbl_num
   quotient = 1
   remainder = 0
   
   While quotient <> 0
      quotient = Math.Floor(dividend / 10)
      remainder = Math.Remainder(dividend, 10)
      
      digits = digits.AddItem(remainder)
      
      dividend = quotient
   Wend
   
   
   fillSpaces = dbl_width - digits.Count
   
   If fillSpaces > 0 Then
      For i = 1 To fillSpaces
         resultStr = resultStr.Append("0")
      Next
      
   EndIf
   
   For i = digits.Count To 1 Step -1
      resultStr = resultStr.Append(digits.GetItemAt(i))
   Next
   
   Return resultStr
EndFunction
```

**Comment by VBAndCs** [comment]
And this is how you can get the integral and decimal parts of a number:
```
X = 10.102
Parts = Text.Split(X, ".", True, False)
TextWindow.WriteLines({
      "integer: " + Parts[1],
      "decimal: " + Parts[2]
})
```
________________________________

**Comment by dynamicboy** [comment]
I didn't realize that numbers could be used as a string.
It will be nice to add a ToString("formatString") method to numbers.

I haven't learn all content of sVB reference book. I just look it up when I'm about to use particular feature.

My Path along with Basic language is: QBasic, QB64, Basic256, Microsoft Small Basic, Small Basic Prime, finally the Small Visual Basic。

I tried to make a few programs with each of them and found out It's hard and time-consuming to present a pretty UI, and difficult to make really useful programs without extensions.

Small Visual Basic is the most convinient one to make pretty UIs. 
The auto-completion makes coding fast and enjoyable.

It's worth promoting more. 

BTW, I found that Sound.Play() is probably playing in background thread, it can play several songs at the same time without making the program waiting. Is there any way to dispatch an operation, such as File.Copy, to a background thread to keep UI active? 
And with Sound, how to get the length of mp3 file in order to know when it stopped.

Thanks for your contribution.
